oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -281,14 +287,72 @@ private HttpResponse getMetadataResponse(String url) throws IOException {
     return response;
   }
 
-  /** Return whether code is running on Google Compute Engine. */
-  static boolean runningOnComputeEngine(
+  /**
+   * Implements an algorithm to detect whether the code is running on Google Compute Environment
+   * (GCE) or equivalent runtime. <a href="https://google.aip.dev/auth/4115">See AIP-4115 for more
+   * details</a> The algorithm consists of active and passive checks: <br>
+   * <b>Active:</b> to check that GCE Metadata service is present by sending a http request to send
+   * a request to {@code ComputeEngineCredentials.DEFAULT_METADATA_SERVER_URL}
+   *
+   * <p><b>Passive:</b> to check if SMBIOS variable is present and contains expected value. This
+   * step is platform specific:
+   *
+   * <p><b>For Linux:</b> check if the file "/sys/class/dmi/id/product_name" exists and contains a
+   * line that starts with Google.
+   *
+   * <p><b>For Windows:</b> to be implemented
+   *
+   * <p>This algorithm overall can be disabled with environment variable {@code
+   * DefaultCredentialsProvider.NO_GCE_CHECK_ENV_VAR} set to true. Returns {@code true} if currently
+   * running on Google Compute Environment (GCE) or equivalent runtime.
+   */
+  static synchronized boolean isOnGce(
       HttpTransportFactory transportFactory, DefaultCredentialsProvider provider) {
     // If the environment has requested that we do no GCE checks, return immediately.
     if (Boolean.parseBoolean(provider.getEnv(DefaultCredentialsProvider.NO_GCE_CHECK_ENV_VAR))) {
       return false;
     }
 
+    boolean result = pingComputeEngineMetadata(transportFactory, provider);
+
+    if (!result) {
+      result = checkStaticGceDetection(provider);
+    }
+
+    LOGGER.log(Level.FINE, "Failed to detect whether running on Google Compute Engine.");
+    return result;
+  }
+
+  @VisibleForTesting
+  static boolean checkProductNameOnLinux(BufferedReader reader) throws IOException {
+    String name = reader.readLine().trim();
+    return name.startsWith(GOOGLE);
+  }
+
+  @VisibleForTesting
+  static boolean checkStaticGceDetection(DefaultCredentialsProvider provider) {
+    String osName = provider.getOsName();
+    try {
+      if (osName.startsWith(LINUX)) {
+        // Checks GCE residency on Linux platform.
+        File linuxFile = new File("/sys/class/dmi/id/product_name");
+        return checkProductNameOnLinux(
+            new BufferedReader(new InputStreamReader(provider.readStream(linuxFile))));
+      } else if (osName.startsWith(WINDOWS)) {
+        // Checks GCE residency on Windows platform.
+        // TODO: implement registry check via FFI
+        return false;
+      }
+    } catch (IOException e) {
+      LOGGER.log(Level.FINE, "Encountered an unexpected exception when checking SMBIOS value", e);
```

Comment:
nit: This log statement should also state what the impact of the failure is:

```suggestion
      LOGGER.log(Level.FINE, "Unable to determine if code is running on GCE. Encountered an unexpected exception when checking SMBIOS.", e);
```

Reply:
The impact depends on environment.. if outside of GCP - it is expected, if inside - either transient error or worse. But we cannot distinguish one from another ourselves if detection fails ) Therefore the log level is fine, not info/warning